agent/router/serf_flooder.go

```diff
@@ -61,49 +56,28 @@ func FloodJoins(logger hclog.Logger, addrFn FloodAddrFn, portFn FloodPortFn,
 			continue
 		}
 
-		// We can't use the port number from the local Serf, so we just
-		// get the host part.
-		addr, _, err := net.SplitHostPort(server.Addr.String())
+		addr, err := addrFn(server)
 		if err != nil {
-			logger.Debug("Failed to flood-join server (bad address)",
-				"server", server.Name,
-				"address", server.Addr.String(),
+			logger.Debug("Failed to flood-join server", "server",
+				server.Name, "address", server.Addr.String(),
 				"error", err,
 			)
-		}
-		if addrFn != nil {
-			if a, ok := addrFn(server); ok {
-				addr = a
-			}
+			continue
 		}
 
-		// Let the callback see if it can get the port number, otherwise
-		// leave it blank to behave as if we just supplied an address.
-		if port, ok := portFn(server); ok {
-			addr = net.JoinHostPort(addr, fmt.Sprintf("%d", port))
-		} else {
-			// If we have an IPv6 address, we should add brackets,
-			// single globalSerf.Join expects that.
-			if ip := net.ParseIP(addr); ip != nil {
-				if ip.To4() == nil {
-					addr = fmt.Sprintf("[%s]", addr)
-				}
-			} else {
-				logger.Debug("Failed to parse IP", "ip", addr)
-			}
-		}
+		dstServerName := fmt.Sprintf("%s.%s", server.Name, server.Datacenter)
 
 		// Do the join!
-		n, err := globalSerf.Join([]string{addr}, true)
+		n, err := dstSerf.Join([]string{dstServerName + "/" + addr}, true)
```

Comment:
I am wondering if this is correct. I think it needs the memberlist and serf changes cherry picked too....

Reply:
Memberlist has been updated already, pulled in the changes from releases/1.7.x

Reply:
👍 If those memberlist changes are in this should be fine.